Solution

panicua opened the pull request.

panicua commented on app/main.py:
```diff
@@ -1,3 +1,122 @@
+    def __repr__(self) -> str:
+        return f"Distance(km={self.km})"
+
+    def __add__(
+        self, other: Union["Distance", int, float]
+    ) -> Union["Distance", TypeError]:
```
I don't know how it works with newer version of Python but on my 3.11 it didn't let me write ```other: Distance``` only ```other: "Distance"``` and after that if I wanted to use ```other: "Distance" | int | float``` tests were giving me an error, that's why I had to use ```Union```.

DmytroHlazyrin replied:
You can use import annotation from __future__ in this case

DmytroHlazyrin commented on app/main.py:
```diff
@@ -1,3 +1,122 @@
+        if isinstance(other, (int, float)):
+            other = Distance(other)
+            return Distance(self.km + other.km)
+        raise TypeError(
+            f"unsupported type for __add__ method. 'Distance', "
+            f"<class 'float'> or <class 'int'> were expected. "
+            f"But received {type(other)}"
+        )
```
if you had used type annotation from future you wouldn't have had to describe the type error in all this cases, as it would be pretty much the same by default.

DmytroHlazyrin reviewed (changes requested):
use import annotation from --future-- and refactor your code

panicua commented:
@DmytroHlazyrin I changed from ```typing > Unite``` to ```__future__ > annotations``` but I left my raise errors cause you can see that future for example for ```__add__``` magic method still doesn't indicate exactly at the problem. (screenshots below, 1st without my error raise, 2nd with my own error raise)
![NVIDIA_Share_xSIw02vWNQ](https://github.com/mate-academy/py-distance-class/assets/13889031/a9a6276f-d828-444a-8f7a-f83d4cf8f228)
![NVIDIA_Share_5w0bZA9OHo](https://github.com/mate-academy/py-distance-class/assets/13889031/19b66082-60d9-4586-b8ad-6d76abe801cd)

YuriiKindrat commented on app/main.py:
```diff
@@ -1,3 +1,119 @@
+            f"But received {type(other)}"
+        )
+
+    def __le__(self, other: int | float) -> bool | TypeError:
+        if isinstance(other, Distance):
+            return self.km <= other.km
+        if isinstance(other, (int, float)):
+            other = Distance(other)
+            return self.km <= other.km
+        raise TypeError(
+            f"unsupported type for __le__ method. <class 'float'> "
+            f"or <class 'int'> were expected. "
+            f"But received {type(other)}"
+        )
+
+    def __ge__(self, other: int | float) -> bool | TypeError:
+        if isinstance(other, Distance):
+            return self.km >= other.km
+        if isinstance(other, (int, float)):
+            other = Distance(other)
+            return self.km >= other.km
+        raise TypeError(
+            f"unsupported type for __ge__ method. <class 'float'> "
+            f"or <class 'int'> were expected. "
+            f"But received {type(other)}"
+        )
```
u can use previous methods here, and implement them in one line )

panicua replied:
you mean that I should user the ```__lt__``` and ```__gt__``` under ```__le__``` and ```__ge__```? cause I didn't rly understand

panicua commented on app/main.py:
```diff
@@ -94,26 +94,8 @@ def __eq__(self, other: int | float) -> bool | TypeError:
-            f"But received {type(other)}"
-        )
+    def __le__(self, other: int | float | Distance) -> bool | TypeError:
+        return self.__eq__(other) or self.__lt__(other)
```
but in this case won't be able to raise a custom error for ```__le__``` and ```__ge__```, it will raise an error from ```__eq__```,```__lt__ or ```__gt__```

Oleksl888 reviewed (changes requested):
This comment by Yuri haven't been resolved yet. Can you please have a look?

> it would be better to write a custom method for error raising to avoid code duplication in every method also, u could write an decorator with try-except inside it, to optimize ur methods workflow and write less code and conditions

panicua commented:
> This comment by Yuri haven't been resolved yet. Can you please have a look?
> 
> > it would be better to write a custom method for error raising to avoid code duplication in every method also, u could write an decorator with try-except inside it, to optimize ur methods workflow and write less code and conditions

we didn't learn Try/Except methods yet, but I can create a static method for comparison magic methods.

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,98 @@
-    # Write your code here
-    pass
+    def __init__(self, km: int | float) -> None:
+        self.km = km
```
What if user passes wrong data type to our class? Should we have some validation in place?

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,98 @@
+    ) -> bool | callable(TypeError):
+        return not self < other
+
+    @staticmethod
```
1. Since your function doesn't do anything except raising an error with a message you can just raise TypeError in your code with a appropriate message.
2. If you want an error handler with certain logic it is define it in a separate class - no need to mix it with other logic. 
3. Also you have 3 methods that do exactly the same - so you could actually have only 1 with different parameters.
4. I understand that exceptions are beyond current scope of what you have learned - so you can remove this block of code and make it more simple.
5. Let me know if you have any questions

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,98 @@
+        return f"Distance(km={self.km})"
+
+    def __add__(self, other: int | float | Distance) -> Distance | TypeError:
+        if isinstance(other, Distance):
```
Simplify your methods to use less if conditions

panicua commented:
-Removed ```@staticmethod```
-Added ```@property``` decorator to check type of an input value.
-Removed all ```raise``` with ```TypeError``` to the better time when we learn about ```Try/Except```

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,73 @@
+
+    @km.setter
+    def km(self, value: int | float) -> None:
+        if not isinstance(value, (int | float)):
```
well done with this! just need a small fix to the second argument to isinstance method. (how do we separate items in tuple?)

panicua replied:
True, totally missed it, already fixed, will push with the next commit.

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,73 @@
+        return f"Distance(km={self.km})"
+
+    def __add__(self, other: int | float | Distance) -> Distance:
+        if isinstance(other, (Distance, int, float)):
```
you basically check for int and float in isinstance method twice. You can simplify this

panicua replied:
I don't get how I can do it with 1 ```return``` and 1 ```if``` cause if I get other as ```Distance``` it's gonna work fine, if I get other as ```int``` or ```float```, I'll have to make ```other = Distance(other)``` so I could compare their ```.km``` after.

Oleksl888 replied:
`    if isinstance(other, (int, float)):
        other = Distance(other)
    return Distance(self.km + other.km)` try following this pattern in your methods to simplify code

panicua replied:
> ` if isinstance(other, (int, float)): other = Distance(other) return Distance(self.km + other.km)` try following this pattern in your methods to simplify code

oh, right I didn't see it at first ;d

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,73 @@
+            return Distance(self.km + other.km)
+
+    def __iadd__(self, other: int | float | Distance) -> Distance:
+        if isinstance(other, (Distance, int, float)):
```
same as above - no need to check the same types twice + no need for nested if's. Also refactor this in other methods.

panicua replied:
I don't get how I can do it with 1 return and 1 if cause if I get other as Distance it's gonna work fine, if I get other as int or float, I'll have to make other = Distance(other) so I could compare their .km after.

Oleksl888 replied:
You can do it, I believe in you :)

Oleksl888 commented on app/main.py:
```diff
@@ -1,3 +1,73 @@
+                other = Distance(other)
+            return self.km == other.km
+
+    def __le__(self, other: int | float | Distance) -> bool:
```
Here you compare self (instance of Distance class) with possible int or float. Also why do you use 'not' operator here?

panicua replied:
> Here you compare self (instance of Distance class) with possible int or float. Also why do you use 'not' operator here?
---Here you compare self (instance of Distance class) with possible int or float.
Actually I believe that's not true, I use ```return not self > other``` so it goes to ```__gt__``` method at this line and ```__gt__``` method is checking if other is ```int```, ```float``` or ```Distance```. If it's ```int``` or ```float``` it does ```other = Distance(other)``` under ```__gt__``` method as you can see. 
p.s. screenshot is attached with comparison of ```Distance``` with ```int/float/Distance```
![NVIDIA_Share_wStengcfP4](https://github.com/mate-academy/py-distance-class/assets/13889031/1a847189-5ef9-45d1-8f8e-9b0c40dd8c06)

---Also why do you use 'not' operator here?
Because when I use ```return not self > other``` in ```__le__``` ```less-equal``` magic method it returns ```True``` when ```self == other``` or ```self < other```, why should I use 2 comparisons when I can just write the key word ```not``` and just use ```self > other```? (Btw this was the reccomendation from @YuriiKindrat on our QnA I had: ```self == other``` or ```self < other``` before that.

Oleksl888 replied:
That's a good point and clever implementation. I totally overlooked that.
Also along with that you can use the following:
[https://docs.python.org/3/library/functools.html#functools.total_ordering](url)

YuriiKindrat commented on app/main.py:
```diff
@@ -1,3 +1,68 @@
+    def __init__(self, km: int | float) -> None:
+        self.km = km
+
+    @property
+    def km(self) -> int | float:
+        return self._km
+
+    @km.setter
+    def km(self, value: int | float) -> None:
+        if not isinstance(value, (int, float)):
+            raise ValueError(
+                f"Unsupported type. <class 'float'> "
+                f"or <class 'int'> were expected. "
+                f"Got {type(value)} instead."
+            )
+        self._km = value
```
GJ!

YuriiKindrat reviewed (approved):
LGTM!
